src/Microsoft.DotNet.CodeFormatter.Analyzers/ExplicitVariableTypeAnalyzer.cs

```diff
@@ -0,0 +1,116 @@
+                                                                            ResourceHelper.MakeLocalizableString(nameof(Resources.ExplicitVariableTypeAnalyzer_MessageFormat)),
+                                                                            "Style",
+                                                                            DiagnosticSeverity.Warning,
+                                                                            true,
+                                                                            customTags: ForEachStatementCustomTag);
+
+        private static readonly ImmutableArray<DiagnosticDescriptor> s_supportedRules = ImmutableArray.Create(s_ruleVariableDeclaration, s_ruleForEachStatement);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => s_supportedRules;
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(syntaxContext =>
+            {
+                var node = (VariableDeclarationSyntax)syntaxContext.Node;
+                // Implicit typed variables cannot have multiple declartors  
+                if (node.Type.IsVar &&
+                    !IsTypeObvious(node) &&
+                    !IsAnonymousType(node.Variables.Single(), syntaxContext.SemanticModel, syntaxContext.CancellationToken))
+                {
+                    syntaxContext.ReportDiagnostic(Diagnostic.Create(s_ruleVariableDeclaration, node.GetLocation(), node.Variables.Single().Identifier.Text));
+                }
+            }, SyntaxKind.VariableDeclaration);
+
+            context.RegisterSyntaxNodeAction(syntaxContext =>
+            {
+                var node = (ForEachStatementSyntax)syntaxContext.Node; 
+                if (node.Type.IsVar &&
+                    !IsTypeObvious(node) &&
+                    !IsAnonymousType(node, syntaxContext.SemanticModel, syntaxContext.CancellationToken))
+                {
+                    syntaxContext.ReportDiagnostic(Diagnostic.Create(s_ruleForEachStatement, node.Identifier.GetLocation(), node.Identifier.Text));
+                }
+            }, SyntaxKind.ForEachStatement);
+        }
+
+        // Return true if given SyntaxNode has an anonymous type, which can't be replaced by an explicit type.
+        private static bool IsAnonymousType(SyntaxNode node, SemanticModel model, CancellationToken cancellationToken)
+        {
+            ISymbol symbol = model.GetDeclaredSymbol(node, cancellationToken); 
+            return ((ILocalSymbol)symbol).Type.IsAnonymousType;
+        }
+
+        // Return true if given SyntaxNode is either VariableDeclarationSyntax or ForEachStatementSyntax and 
+        // VariableDeclarationSyntax.Variables.Single().Initializer.Value or ForEachStatementSyntax.Expression is:
```

Comment:
This puts a lot of implementation into the comment. I'd replace the text on this line (that is, the text following the "and" on the previous line) with:

"... the variable is initialized with one of the following:"